Userland/Libraries/LibSemVer/SemVer.cpp

```diff
@@ -0,0 +1,305 @@
+    if (version.count('+') > 1)
+        return AK::Error::from_string_view("Build metadata must be defined once"sv);
+
+    GenericLexer lexer(version.trim_whitespace());
+    if (lexer.tell_remaining() == 0)
+        return AK::Error::from_string_view("Version string is empty"sv);
+
+    // Parse the normal version parts
+    // https://semver.org/#spec-item-2
+    auto version_part = lexer.consume_until(m_number_separator).to_uint<u64>();
+    if (!version_part.has_value())
+        return AK::Error::from_string_view("Major version is not numeric"sv);
+    auto v_major = version_part.value();
+
+    lexer.consume();
+
+    version_part = lexer.consume_until(m_number_separator).to_uint<u64>();
+    if (!version_part.has_value())
+        return AK::Error::from_string_view("Minor version is not numeric"sv);
+    auto v_minor = version_part.value();
+
+    lexer.consume();
+
+    version_part = lexer.consume_while<Function<bool(char)>>([](char ch) {
+                            return ch >= '0' && ch <= '9';
+                        })
+                       .to_uint<u64>();
+    if (!version_part.has_value())
+        return AK::Error::from_string_view("Patch version is not numeric"sv);
+    auto v_patch = version_part.value();
+
+    if (lexer.is_eof())
+        return SemVer(v_major, v_minor, v_patch, m_number_separator);
+
+    String buildmetadata;
+    Vector<String> prerelease_identifiers;
+
+    // Note: This is efficient that containsa
+    auto constexpr has_badchars = [](String const& haystack) -> bool {
+        for (const auto ch : haystack.bytes()) {
+            if (ch == ' ' || ch == '_')
+                return true;
+        }
+        return false;
+    };
+
+    switch (lexer.consume()) /* just after patch ch */
+    {
+    case '+': {
+        // Spec: https://semver.org/#spec-item-10
+        if (lexer.is_eof())
+            return AK::Error::from_string_view("Buildmeta can't be empty"sv);
+
+        buildmetadata = TRY(TRY(String::from_utf8(lexer.consume_all())).trim_ascii_whitespace());
+        if (buildmetadata.starts_with('.'))
+            return AK::Error::from_string_view("Invalid build metadata part"sv);
+        if (has_badchars(buildmetadata))
+            return AK::Error::from_string_view("Buildmeta can't contain bad characters"sv);
+
+        break;
+    }
+    case '-': {
+        // https://semver.org/#spec-item-9
+        auto prerelease = TRY(TRY(String::from_utf8(lexer.consume_until('+'))).trim_ascii_whitespace());
+
+        // Check for empty prerelese given hyphen (-) scanned
+        if (prerelease.is_empty())
+            return AK::Error::from_string_view("Prerelease can't be empty"sv);
+        // Check for bad characters in the substring
+        if (has_badchars(buildmetadata))
+            return AK::Error::from_string_view("Prerelease can't contain bad characters"sv);
+
+        // Validate prerelease against charset
+        Regex<ECMA262Parser> const identifier_pattern("^[A-Za-z0-9-\\.]+$");
```

Comment:
Regex seems a bit overkill for this (especially if it remains as the only usage), maybe we should just use `code_points()` on the identifier and iterate over it?